# <center>Homework 2 Intro to Tensorflow</center>


# <center>COSC 4337</center>


# <center>Dr. Rizk</center>

# <center>Dosbol Aliev</center>

# <center>MyUH 1867424</center>

### Introduction


Until now, you may have used sklearn or numpy to build neural networks. Now we will
step you through a deep learning framework that will allow you to build neural networks
more easily. Machine learning frameworks like TensorFlow, Torch, Caffe, Keras, and
many others can speed up your machine learning development significantly. All of these
frameworks also have a lot of documentation, which you should feel free to read. In this
assignment, you will learn to do the following in TensorFlow:<br>
    
● Initialize variables<br>
● Perform operations with tensors<br>
● Implement a Neural Network<br>
Programing frameworks can not only shorten your coding time, but sometimes also<br>
perform optimizations that speed up your code.

# <center>1 - Exploring the Tensorflow Library</center>

To begin, you will need the following imports

In [1]:
# Importing all needed libraries to our Jupiter Notebook
import math
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

import warnings
warnings.filterwarnings("ignore") # Ignoring Warnings
np.random.seed(1)

### 1 - Exploring the Tensorflow Library Cont...

Now that you have imported the library, we will walk you through its different
applications. You will start with an example, where we compute for you the loss of
one training example

Run the following lines of code. Given the formula above, does the output make
sense?

In [2]:
y_hat=tf.constant(36) #Define y_hat constant. Set to 36
y=tf.constant(39) # Define y. Set to 39

loss=(y-y_hat)**2 #Calculate the loss
tf.print(loss)

9


### 1 - Exploring the Tensorflow Library Cont...

Note: Most functions that you could apply to a numpy array will also work on
tensorflow tensors, but instead of np.function() it will be tf.function()

### 1.1 Initializing Tensors

Let's see some more examples of how we could initialize tensors. Run the
following lines of code on your own and see what the output is. This should look
very familiar if you’re used to numpy

In [3]:
x=tf.constant(4,shape=(1,1), dtype=tf.float32)
tf.print(x) # We are creating matrix of 4

[[4]]


In [4]:
x=tf.constant([[1,2,3],[4,5,6]])
tf.print(x) # Since we have 2 matrices and they have the same number of rows, with tf.constant, 
# we are creating a new row to make a second matrix at the bottom of the first one

[[1 2 3]
 [4 5 6]]


In [5]:
x=tf.ones((3,3))
tf.print(x) # We are creating a matrix of 3x3 full of 1's(each number is equal to 1)

[[1 1 1]
 [1 1 1]
 [1 1 1]]


In [6]:
x=tf.zeros((3,3))
tf.print(x) # We are creating a matrix of 3x3 full of 0's(each number is equal to 0)

[[0 0 0]
 [0 0 0]
 [0 0 0]]


In [7]:
x=tf.eye(3)
tf.print(x) # We are creating a matrix by using tf.eye of 3x3 matrix and making diagonal numbers equal to 1

[[1 0 0]
 [0 1 0]
 [0 0 1]]


### 1.2 Mathematical Operations


Now let's see some examples of mathematical operations that that we can
perform with tensors. Run the following lines of code and see what the output is.

In [8]:
x=tf.constant([1,2,3])
y=tf.constant([9,8,7])

z=x+y
tf.print(z) # We are adding 1st-row numbers and 2nd-row numbers. For example, 1+9=10, 2+8=10, 3+7=10

[10 10 10]


In [9]:
z=x-y
tf.print(z) # We are subtracting 1st-row numbers from 2nd-row numbers. For example, 1-9=-8, 2-8=-6, 3-7=-4

[-8 -6 -4]


In [10]:
z=x/y
tf.print(z) # We are dividing 1st-row numbers into 2nd-row numbers. For example, 1/9=0.1111111111111111, 2/8=0.25, 3/7=0.42857142857142855

[0.1111111111111111 0.25 0.42857142857142855]


In [11]:
z=x*y
tf.print(z)# We are multiplying 1st-row numbers into 2nd-row numbers. For example, 1*9=9, 2*8=16, 3*7=21

[9 16 21]


In [12]:
z=tf.tensordot(x,y,axes=1)
tf.print(z) # So, here we multiply first and then add them together. For example
# (1*9)=9
# (2*8)=16
# (3*7)=21
#and then just adding them together 9+16+21= 46

46


In [13]:
z=x**5
tf.print(z) #---> x=tf.constant([1,2,3]). So, basically, we are powering numbers by 5. For example, 1^5=1, 2^5=32, 3^5=243

[1 32 243]


In [14]:
x=tf.random.normal((2,3)) # We are randomized X
y=tf.random.normal((3,4)) # We are randomized Y

z=tf.matmul(x,y) # We are multiplying X matrix to Y matrix using special keyboard MATMUL
tf.print(z)

[[-4.14169121 3.72763848 3.07543683 -1.22759295]
 [2.54111814 -2.38830709 -2.22173452 0.809899926]]


### 1.3 Indexing


To see how we can index tensors, run the following lines of code.

In [15]:
x=tf.constant([0,1,1,2,3,1,2,3])
print(x[:]) #We are getting all numbers with no expectations!

tf.Tensor([0 1 1 2 3 1 2 3], shape=(8,), dtype=int32)


In [16]:
print(x[1:]) #We are getting all numbers. EXPECT the first number whose index is 0 (number=0)

tf.Tensor([1 1 2 3 1 2 3], shape=(7,), dtype=int32)


In [17]:
print(x[1:3]) #We are getting index of number between 1 and 3

tf.Tensor([1 1], shape=(2,), dtype=int32)


In [18]:
print(x[::2]) #We are getting even index of numbers (index: 0,2,4,6)

tf.Tensor([0 1 3 2], shape=(4,), dtype=int32)


In [19]:
print(x[::-1]) #We are reversing the matrix. So, now it starts backward.

tf.Tensor([3 2 1 3 2 1 1 0], shape=(8,), dtype=int32)


In [20]:
indices =tf.constant([0,3])
x_ind=tf.gather(x,indices)
print(x_ind)# It gives us the numbers' index in x. 
#For Example, we have 0 and 3 and when will look at x matrix we can see index for number 0 is 0 and the index of number 3 is 2

tf.Tensor([0 2], shape=(2,), dtype=int32)


In [21]:
x=tf.constant([[1,2],[3,4],[5,6],])
print(x[0,:]) # We are choosing the first index of the matrix which is [1,2](a subset of the big matrix)

tf.Tensor([1 2], shape=(2,), dtype=int32)


In [22]:
print(x[0:2,:]) # We are printing all index until index 2 so our first index(index=0) is [1,2] and second index(index=1) is [3,4]
# and the shape says size should be 2 x2, so we are just making the second matrix at the bottom of the first one

tf.Tensor(
[[1 2]
 [3 4]], shape=(2, 2), dtype=int32)


### 1.4 Reshaping

Now let's see how we can reshape. This is something you'll do a lot when passing
input into your models. Run the following lines of code to see what happens.

In [23]:
x=tf.range(9)
print(x) #We are creating a range of numbers from 0 to 8, not including itself(number =9)

tf.Tensor([0 1 2 3 4 5 6 7 8], shape=(9,), dtype=int32)


In [24]:
x=tf.reshape(x,(3,3))
print(x) # Reshaping means that we are now reshaping the previous matrix, which was only in 1 row, to 3x3 matrix

tf.Tensor(
[[0 1 2]
 [3 4 5]
 [6 7 8]], shape=(3, 3), dtype=int32)


In [25]:
x=tf.transpose(x,perm=[1,0])
print(x) # Transpose means it will change places of rows and columns. For example, in the first row, we have 0,1,2 after transposing
# it will become  0
            #     1
            #     2

tf.Tensor(
[[0 3 6]
 [1 4 7]
 [2 5 8]], shape=(3, 3), dtype=int32)


### Exercise


Let's start this programming exercise by computing the following equation:
𝑌=𝑊𝑋+𝑏 , where 𝑊 and 𝑋 are random matrices and b is a random vector.
<br>
Exercise: Compute 𝑊𝑋+𝑏 where 𝑊,𝑋 , and 𝑏 are drawn from a random normal
distribution. W is of shape (4, 3), X is (3,1) and b is (4,1).
<br>
You might find the following functions helpful:
    <br><br>
● tf.matmul(..., ...) to do a matrix multiplication<br>
● tf.add(..., ...) to do an addition<br>
● np.random.randn(...) to initialize randomly or tensorflow random<br>
<br>
See the next slide for how you should create this function

In [26]:
def linear_function():
    np.random.seed(1)
    
    X = tf.constant(np.random.randn(3, 1), name="X") # First, we need to Initializes X to be a random tensor of shape (3,1)
    W = tf.constant(np.random.randn(4, 3), name="W") # Second, we need to Initializes W to be a random tensor of shape (4,3)
    b = tf.constant(np.random.randn(4, 1), name="b") # Third, we need to Initializes b to be a random tensor of shape (4,1)
    Y = tf.add(tf.matmul(W, X), b) # Then we first using matmul to multiple W*X and the we are using add function to add b
    # So our formula is Y = W * X + b
    
    
    return Y # We are calling Y variable
linear_function()

<tf.Tensor: shape=(4, 1), dtype=float64, numpy=
array([[-2.15657382],
       [ 2.95891446],
       [-1.08926781],
       [-0.84538042]])>

# <center>2 - Building your first neural network in tensorflow</center>


For this section, we’ll be showing you how to construct a neural network. Please
refer back to lecture slides if you’re still iffy or rusty on how neural networks work
and things like forward, backpropagation, cost functions, etc. To begin, we’ll need
the following imports.

In [27]:
# Importing needed Libraries
from tensorflow import keras 
from tensorflow.keras import layers
from tensorflow.keras.datasets import mnist

### 2 - Building your first neural network in tensorflow Cont...


For this section we'll be working with the MNIST dataset. This dataset contains
images of digits from 0 to 9. The images are grayscale meaning they only have
one channel and the pixels are 28 x 28. Our goal will be to classify each image

Here we can see an
example of the MNIST
dataset and the type of
images that we’ll be working
with.

### 2.1 - Loading the data


Run the following lines of code

In [28]:
(x_train,y_train),(x_test,y_test) = mnist.load_data()
print(x_train.shape) # We are training and test X and y

(60000, 28, 28)


After loading the data and printing the shape of x_train, we can see that there are
60,000 images, where each image is 28 x 28

#### Recall the general form of a neural network. Can we pass a whole image into this?

Yes, we can pass a whole image into this

Since we're going to send this data into a neural network, we must flatten the data
so that we have 1 long column with the feature values. In other words, if you recall
how input is passed into the neural network, we'll have to make sure that all
60,000 images are flattened into 1 long column. We'll later look into convolutional
neural networks, which are much better at dealing with image data.

Run the following lines of code to accomplish this

Note, the -1 ensures that we keep the 1st dimension the same (60,000) because
what we want to do is flatten each of the 60,000 images. And converting to float an
normalizing (dividing by 255) are just so that we can have faster computations

In [29]:
x_train = x_train.reshape(-1,28*28).astype("float32")/255.0
x_test = x_test.reshape(-1,28*28).astype("float32")/255.0 # Explanations are above this line

## Creating Our Simple Neural Net

To start, we’ll create our simple neural network using the “sequential “ from Keras.
The sequential is very convenient, but lacks flexibility. See the following slides for
more details

### Do’s and Dont’s of Sequential

When to use a Sequential model?

A Sequential model is appropriate for a plain stack of layers where each layer has
exactly one input tensor and one output tensor.


A Sequential model is not appropriate when:
    
● Your model has multiple inputs or multiple outputs <br>
● Any of your layers has multiple inputs or multiple outputs <br>
● You need to do layer sharing <br>
● You want non-linear topology (e.g. a residual connection, a multi-branch
model)

### Creating a Sequential model

You can create a Sequential model by passing a list of layers to the Sequential
constructor:

In [30]:
model = keras.Sequential ( [
    layers.Dense(2,activation="relu"),
    layers.Dense(3,activation='relu'),
    layers.Dense(4),
])

### Creating a Sequential model Continued...

Its layers are accessible via the layers attribute:

In [31]:
model.layers

### Sequential for our neural network

#### Run the following lines of code and answer the questions on the next slide

In [32]:
#creatunf our layers
model = keras.Sequential([
    
    layers.Dense(200, activation='relu'),  #<---- FULLY connected layer
    layers.Dense(100, activation="relu"),
    layers.Dense(10),
])
#telling keras how to configure the training part of our network. For example 
# What loss do you want to use for training etc
model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer = keras.optimizers.Adam(lr=0.00001),
    metrics=['accuracy'],
)

#fit/train our model. similar to sklearn's fit
model.fit(x_train,y_train,batch_size=32,epochs=1,verbose=2)

#evaluate using the test data

model.evaluate(x_test,y_test,batch_size=32,verbose=2)

1875/1875 - 7s - loss: 1.5921 - accuracy: 0.6097 - 7s/epoch - 4ms/step
313/313 - 1s - loss: 0.8887 - accuracy: 0.8330 - 1s/epoch - 4ms/step


[0.8886922597885132, 0.8330000042915344]

# <center> Questions:</center>


#### 1.Why is the last fully connected layer of size 10?<br>
It is 10 because there are 10 digits we are trying to classify<br>
#### 2.Is cross entropy a good loss function for our problem?<br>
In our case, cross-entropy is a suitable loss function because it is frequently employed for "optimizing classification models."

#### 3.How can you get the model to perform greater than 95% accuracy on our images? Rewrite the model in a new cell.

In [33]:
#creatunf our layers
model = keras.Sequential([
    
    layers.Dense(200, activation='relu'), #We are creating 200 layers #<---- FULLY connected layer
    layers.Dense(100, activation="relu"),#We are creating 100 layers
    layers.Dense(10),#We are creating 10 layers
])
#telling keras how to configure the training part of our network. For example 
# What loss do you want to use for training etc
model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer = keras.optimizers.Adam(lr=0.001),
    metrics=['accuracy'],
)

#fit/train our model. similar to sklearn's fit
model.fit(x_train,y_train,batch_size=32,epochs=10,verbose=2)

#evaluate using the test data

model.evaluate(x_test,y_test,batch_size=32,verbose=2)

Epoch 1/10
1875/1875 - 7s - loss: 0.2140 - accuracy: 0.9372 - 7s/epoch - 4ms/step
Epoch 2/10
1875/1875 - 7s - loss: 0.0893 - accuracy: 0.9732 - 7s/epoch - 4ms/step
Epoch 3/10
1875/1875 - 6s - loss: 0.0601 - accuracy: 0.9811 - 6s/epoch - 3ms/step
Epoch 4/10
1875/1875 - 7s - loss: 0.0457 - accuracy: 0.9853 - 7s/epoch - 4ms/step
Epoch 5/10
1875/1875 - 6s - loss: 0.0374 - accuracy: 0.9880 - 6s/epoch - 3ms/step
Epoch 6/10
1875/1875 - 6s - loss: 0.0282 - accuracy: 0.9906 - 6s/epoch - 3ms/step
Epoch 7/10
1875/1875 - 6s - loss: 0.0255 - accuracy: 0.9914 - 6s/epoch - 3ms/step
Epoch 8/10
1875/1875 - 6s - loss: 0.0219 - accuracy: 0.9925 - 6s/epoch - 3ms/step
Epoch 9/10
1875/1875 - 7s - loss: 0.0182 - accuracy: 0.9939 - 7s/epoch - 4ms/step
Epoch 10/10
1875/1875 - 6s - loss: 0.0149 - accuracy: 0.9955 - 6s/epoch - 3ms/step
313/313 - 1s - loss: 0.0972 - accuracy: 0.9776 - 1s/epoch - 4ms/step


[0.09717962890863419, 0.9775999784469604]

#### So, I used 10 epochs and increased the learning rate to 0.001 (from 0.00001). As We can see, our accuracy rate became higher than 95%(97.7%)

### <center>Using Functional API</center>

In [34]:
inputs =keras.Input(shape=(28*28))
x = layers.Dense(512, activation ='relu')(inputs)#pass in inputs
x = layers.Dense(256, activation ='relu')(x) #pass in previous layer's output
outputs=layers.Dense(10, activation ="softmax")(x)#pass in previus layer's output
model = keras.Model(inputs=inputs,outputs=outputs)

model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=False),
    optimizer = keras.optimizers.Adam(lr=0.001),
    metrics= ['accuracy'],
)

model.fit(x_train,y_train,batch_size=32, epochs=5, verbose=2)
model.evaluate(x_test,y_test,batch_size=32,verbose=2)

Epoch 1/5
1875/1875 - 10s - loss: 0.1851 - accuracy: 0.9441 - 10s/epoch - 5ms/step
Epoch 2/5
1875/1875 - 9s - loss: 0.0790 - accuracy: 0.9757 - 9s/epoch - 5ms/step
Epoch 3/5
1875/1875 - 9s - loss: 0.0532 - accuracy: 0.9834 - 9s/epoch - 5ms/step
Epoch 4/5
1875/1875 - 9s - loss: 0.0433 - accuracy: 0.9869 - 9s/epoch - 5ms/step
Epoch 5/5
1875/1875 - 9s - loss: 0.0324 - accuracy: 0.9893 - 9s/epoch - 5ms/step
313/313 - 1s - loss: 0.0540 - accuracy: 0.9831 - 1s/epoch - 4ms/step


[0.05402835085988045, 0.9830999970436096]

Please run the following and answer the questions on the next slide.

# <center>Questions</center>


#### 1. How is functional different from sequential, and why is it considered more flexible?You may be able to tell by just looking at the code.<br>
In Function API, we are using specific giving the size of images as 28 x 28 and using this as a variable, while in Sequential, we don't assign. Also, we are using "Softmax." Also, the main difference, as I see in "SparseCategoricalCrossentropy(from_logits=True)" in Function API, it is False, while in the Sequential is TRUE, but other things in the "model. compile" everything is the same

#### 2. What accuracy did you obtain, and how can it be improved if at all?
For Sequential, with the number of epochs of 10 and a learning rate of 0.001, I got 97.7% accuracy, while in the Functional API, with only 5 epochs and the same learning rate, I got 98.3% accuracy. As we can see, Functional API gave us higher accuracy. And to improve it we should have more echoes and probably increase the learning rate as well


# <center>Working with outside data</center>

Congrats, you should now know how to build simple neural networks using tensorflow . Note, that the data we used
here came form tensorflow. However, we’ll often have to load data from our own computer or other sources (typically
images or CSV files). As the final part of this homework, please follow along with this excellent source that comes
directly from tensorflow showing you how to work with CSV files.<br>

https://www.tensorflow.org/tutorials/load_data/csv


In [35]:
# Importing all needed Libraries 
import pandas as pd
import numpy as np

# Make numpy values easier to read.
np.set_printoptions(precision=3, suppress=True)

import tensorflow as tf
from tensorflow.keras import layers

In [36]:
# Getting Dataset Online and creating Labels for it
abalone_train = pd.read_csv(
    "https://storage.googleapis.com/download.tensorflow.org/data/abalone_train.csv",
    names=["Length", "Diameter", "Height", "Whole weight", "Shucked weight",
           "Viscera weight", "Shell weight", "Age"])

abalone_train.head()

Length  Diameter  Height  Whole weight  Shucked weight  Viscera weight  \
0   0.435     0.335   0.110         0.334          0.1355          0.0775   
1   0.585     0.450   0.125         0.874          0.3545          0.2075   
2   0.655     0.510   0.160         1.092          0.3960          0.2825   
3   0.545     0.425   0.125         0.768          0.2940          0.1495   
4   0.545     0.420   0.130         0.879          0.3740          0.1695   

   Shell weight  Age  
0        0.0965    7  
1        0.2250    6  
2        0.3700   14  
3        0.2600   16  
4        0.2300   13

In [37]:
abalone_features = abalone_train.copy()
abalone_labels = abalone_features.pop('Age') # We are removing Label called "Age"

In [38]:
abalone_features = np.array(abalone_features) # Putting them to Array list
abalone_features #printing them

array([[0.435, 0.335, 0.11 , ..., 0.136, 0.077, 0.097],
       [0.585, 0.45 , 0.125, ..., 0.354, 0.207, 0.225],
       [0.655, 0.51 , 0.16 , ..., 0.396, 0.282, 0.37 ],
       ...,
       [0.53 , 0.42 , 0.13 , ..., 0.374, 0.167, 0.249],
       [0.395, 0.315, 0.105, ..., 0.118, 0.091, 0.119],
       [0.45 , 0.355, 0.12 , ..., 0.115, 0.067, 0.16 ]])

In [39]:
abalone_model = tf.keras.Sequential([ # Here we are using Sequential 
  layers.Dense(64), # First, making label number 64
  layers.Dense(1) # Then just one
])

abalone_model.compile(loss = tf.keras.losses.MeanSquaredError(), # We are telling here that we want to use MeanSquaredError for our loss function
                      optimizer = tf.keras.optimizers.Adam()) # it is like previosly

In [40]:
abalone_model.fit(abalone_features, abalone_labels, epochs=10)

Epoch 1/10
104/104 [==============================] - 1s 3ms/step - loss: 63.3268
Epoch 2/10
104/104 [==============================] - 0s 3ms/step - loss: 12.2554
Epoch 3/10
104/104 [==============================] - 0s 3ms/step - loss: 8.9297
Epoch 4/10
104/104 [==============================] - 0s 3ms/step - loss: 8.3404
Epoch 5/10
104/104 [==============================] - 0s 3ms/step - loss: 7.8196
Epoch 6/10
104/104 [==============================] - 0s 4ms/step - loss: 7.4004
Epoch 7/10
104/104 [==============================] - 0s 3ms/step - loss: 7.0814
Epoch 8/10
104/104 [==============================] - 0s 4ms/step - loss: 6.8521
Epoch 9/10
104/104 [==============================] - 0s 3ms/step - loss: 6.6833
Epoch 10/10
104/104 [==============================] - 0s 2ms/step - loss: 6.5586


In [41]:
normalize = layers.Normalization() # a layer of preparation that normalizes continuous characteristics(Got from the link which you provided me)

In [42]:
normalize.adapt(abalone_features)(#The data's mean and variance will be computed by adapt() and stored as the layer's weights.)(Got from the link which you provided me)

In [43]:
norm_abalone_model = tf.keras.Sequential([
  normalize,
  layers.Dense(64),
  layers.Dense(1)
])

norm_abalone_model.compile(loss = tf.keras.losses.MeanSquaredError(),# We are telling here that we want to use MeanSquaredError for our loss function
                           optimizer = tf.keras.optimizers.Adam())

norm_abalone_model.fit(abalone_features, abalone_labels, epochs=10)

Epoch 1/10
104/104 [==============================] - 1s 3ms/step - loss: 93.8425
Epoch 2/10
104/104 [==============================] - 0s 2ms/step - loss: 54.7680
Epoch 3/10
104/104 [==============================] - 0s 3ms/step - loss: 16.3701
Epoch 4/10
104/104 [==============================] - 0s 3ms/step - loss: 5.7662
Epoch 5/10
104/104 [==============================] - 0s 3ms/step - loss: 5.0682
Epoch 6/10
104/104 [==============================] - 0s 3ms/step - loss: 4.9980
Epoch 7/10
104/104 [==============================] - 0s 3ms/step - loss: 4.9882
Epoch 8/10
104/104 [==============================] - 0s 3ms/step - loss: 4.9691
Epoch 9/10
104/104 [==============================] - 0s 3ms/step - loss: 4.9514
Epoch 10/10
104/104 [==============================] - 0s 3ms/step - loss: 4.9320
